test/LibraryManager.IntegrationTest/LibmanTestsUtility.cs

```diff
@@ -8,6 +9,164 @@ namespace Microsoft.Web.LibraryManager.IntegrationTest
 {
     public class LibmanTestsUtility
     {
+        private static HashSet<string> GetTopLevelDirectoriesAndFiles(string cwd, bool caseInsensitive)
+        {
+            HashSet<string> topLevelItems = caseInsensitive ? new HashSet<string>(StringComparer.OrdinalIgnoreCase) : new HashSet<string>();
+            var dir = new DirectoryInfo(cwd);
+
+            if(dir.Exists)
+            {
+                foreach(FileSystemInfo item in dir.EnumerateDirectories())
+                {
+                    topLevelItems.Add(item.Name);
+                }
+                foreach(FileSystemInfo item in dir.EnumerateFiles())
+                {
+                    topLevelItems.Add(item.Name);
+                }
+            }
+
+            return topLevelItems;
+        }
+
+        public static void WaitForRestoredFiles(string cwd, IEnumerable<string> expectedFilesAndFolders, bool caseInsensitive, int timeout = 10000)
+        {
+            string errorMessage = WaitForRestoredFilesHelper(cwd, expectedFilesAndFolders, caseInsensitive, timeout);
+
+            if (errorMessage != null)
+            {
+                string newErrorMessage = WaitForRestoredFilesHelper(cwd, expectedFilesAndFolders, caseInsensitive, timeout);
+
+                if (newErrorMessage != null)
+                {
+                    errorMessage = String.Concat(errorMessage, "\r\nFailed even when forcing completion with double timeout");
+                }
+                else
+                {
+                    errorMessage = String.Concat(errorMessage, "\r\n*Didn't* fail when forcing completion with double timeout");
+                }
+
```

Comment:
If we're going to do the "doubled timeout" thing in multiple places, it should be its own helper.

Reply:
For here. I use the doubled timeout for WaitForRestoredFile for file restore. But not using timeout for WaitForRestoredFileNotPresent because it is called after the WaitForRestore just verifying the not specified files are not included. So I think this doubled timeout can be written here.

---
In reply to: [194547985](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194547985) [](ancestors = 194547985)

Reply:
It's now only used here for the waitforrestoredfiles.

---
In reply to: [194883862](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194883862) [](ancestors = 194883862,194547985)

Reply:
The point is that we're copying the logic of retrying with an extra timeout duration from the completion helper method.  Is this a pattern we want to do commonly?  Then there should be a base waiter method to contain that retry logic.

Reply:
I see. using a delagate waiter for both waitforrestoredfileshelper and waitfordeletedfileshelper function 

---
In reply to: [195513255](https://github.com/aspnet/LibraryManager/pull/113#discussion_r195513255) [](ancestors = 195513255)